src/coreclr/vm/interpexec.cpp

```diff
@@ -291,6 +293,79 @@ template <typename TResult, typename TSource> void ConvOvfHelper(int8_t *stack,
     }
 }
 
+void* DoGenericLookup(void* genericVarAsPtr, InterpGenericLookup* pLookup)
+{
+    // TODO! If this becomes a performance bottleneck, we could expand out the various permutations of this
+    // so that we have 24 versions of lookup (or 48 is we allow for avoiding the null check), do the only 
+    // if check to figure out which one to use, and then have the rest of the logic be straight-line code.
+    MethodTable *pMT = nullptr;
+    MethodDesc* pMD = nullptr;
+    uint8_t* lookup;
+    if (pLookup->lookupType == InterpGenericLookupType::This)
+    {
+        OBJECTREF thisPtr = ObjectToOBJECTREF((Object*)genericVarAsPtr);
+        NULL_CHECK(thisPtr);
+        pMT = thisPtr->GetMethodTable();
+        lookup = (uint8_t*)pMT;
+    }
+    else if (pLookup->lookupType == InterpGenericLookupType::Class)
+    {
+        pMT = (MethodTable*)genericVarAsPtr;
+        lookup = (uint8_t*)pMT;
+    }
+    else
+    {
+        assert(pLookup->lookupType == InterpGenericLookupType::Method);
+        pMD = (MethodDesc*)genericVarAsPtr;
+        lookup = (uint8_t*)pMD;
+    }
+    void* result = 0;
+
+    uint16_t indirections = pLookup->indirections;
+    if (indirections == 0)
+    {
+        return lookup;
+    }
+    else if (indirections != InterpGenericLookup_UseHelper)
+    {
+        lookup = VolatileLoadWithoutBarrier((uint8_t**)(lookup + pLookup->offsets[0]));
+        if (indirections >= 3)
+            lookup = VolatileLoadWithoutBarrier((uint8_t**)(lookup + pLookup->offsets[1]));
+        if (indirections >= 4)
+            lookup = VolatileLoadWithoutBarrier((uint8_t**)(lookup + pLookup->offsets[2]));
+        do {
+            size_t lastOffset = pLookup->offsets[indirections - 1];
+            if (pLookup->sizeOffset != CORINFO_NO_SIZE_CHECK)
+            {
+                /* Last indirection to check is the size*/
+                size_t size = VolatileLoadWithoutBarrier((size_t*)(lookup + pLookup->sizeOffset));
+                if (size <= lastOffset)
+                {
+                    result = GenericHandleCommon(pMD, pMT, pLookup->signature);
+                    break;
+                }
+            }
+            lookup = VolatileLoadWithoutBarrier((uint8_t**)(lookup + lastOffset));
+
+            if (lookup == NULL)
+            {
+                result = GenericHandleCommon(pMD, pMT, pLookup->signature);
+                break;
+            }
+            result = lookup;
+        } while (0);
+    }
+    else
+    {
+        return GenericHandleCommon(pMD, pMT, pLookup->signature);
+    }
+    return result;
+}
+
+#define DO_GENERIC_LOOKUP(genericVar, lookupIndex) \
```

Comment:
Do we even need this define anymore ? Doesn't seem to add more clarity. We could just replace it with a method returning result explicitly